pallene/checker.lua

```diff
@@ -549,19 +598,43 @@ function Checker:check_stat(stat)
     elseif tag == "ast.Stat.If" then
         stat.condition = self:check_exp_synthesize(stat.condition)
         check_type_is_condition(stat.condition, "if statement condition")
-        self:check_stat(stat.then_)
-        self:check_stat(stat.else_)
+        self:check_stat(stat.then_, false)
+        self:check_stat(stat.else_, false)
 
     elseif tag == "ast.Stat.Break" then
         -- ok
 
+    elseif tag == "ast.Stat.Func" then
+
+        local decl = stat.decl
+        local fname_exp = stat.name
+        decl._type = self:from_ast_type(decl.type)
+        self:check_funcname(fname_exp, decl)
+        stat.value =
+            self:check_exp_verify(stat.value, decl._type, "toplevel function")
+
     else
         typedecl.tag_error(tag)
     end
 
     return stat
 end
 
+function Checker:check_funcname(name, decl)
+    local var = name.var
+    local tag = var._tag
+    decl.name = var.name
+    if tag == "ast.Var.Dot" then
+        local mod_cname = self.symbol_table:find_symbol(var.exp.var.name)
```

Comment:
Be careful when you access a deep sequence of fields like `var.exp.var.name`. We need to check that the `var.exp` is an `ast.Exp.Var` before we can access the `var.exp.var`. We need to check that this last var is an `ast.Var.Name` before we read the `.name` field.

As a rule of thumb, before we use any datatype that is a tagged union, we need to check the `_tag` with either an if statement or an `assert`. Whichever is more appropriate.